### ISSUE TITLE
B007 does not detect shadowed variables

### ISSUE THREAD

**Issue by LordAro** [astral-sh/ruff#14113]
B007 (unused-loop-control-variable) does not trigger when variables are shadowed.

See this extremely contrived example:

```python3
my_dict = my_other_dict = {}
my_list = []
for k, (v1, v2) in my_dict.items():
    for v1 in my_other_dict.values():
        print(v1)
    v2 = "flibble"
    print([k for k in my_list if k > 0])
```

https://play.ruff.rs/92ded63b-80d4-4d7d-989a-3460865af51b

(Example is somewhat nested loop focused as that's what I thought the problem was initially)

Interestingly, PERF102 *does* trigger with this example, but I think only because it doesn't process comprehensions at all? If it's expanded out the warning goes away.

The only thing of note from ALL is PLW2901, which detects the variable shadowing.

Related issues: #8786 #9454 #10723

ruff 0.7.2

**Comment by dylwil3** [comment]
I guess technically my previous example wasn't "shadowing", but an issue of scope. But shadowing _also_ doesn't work (as you showed well in your original example):

```python
for k in ls:
    k = 2
```

But, as you also point out, this is caught by [PLW2901](https://docs.astral.sh/ruff/rules/redefined-loop-name/).
